### PR TITLE
Store comments as entities in Datastore

### PR DESCRIPTION
**Pull request by chenzizhao**
Comments stored locally in an `ArrayList` are erased when the server restarts or scales. A safer, more persistent way to store the data is in a database (`Datastore` service in this case). To achieve this, two files are touched: 

- pom.xml: add google.appengine dependency to use `Datastore` service. 
- DataServlet.java: create, store and load comments as datastore entities. 

(Week 3 Step 5)

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -40,7 +52,9 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String newComment = request.getParameter("new-comment");
     if (!newComment.isEmpty()){
-      this.comments.add(newComment);
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("content", newComment);
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:56]: I recommend you also add a timestamp property similar to the examples. This way you can show the comments in order . Typically when you read from a database, order is not guaranteed so its a good idea to store time as well.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -26,12 +32,18 @@
...
 public class DataServlet extends HttpServlet {
-  private ArrayList<String> comments = new ArrayList<String>( 
-    Arrays.asList());
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
   
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = new Gson().toJson(this.comments);
+    ArrayList<String> comments = new ArrayList<>();
+    Query query = new Query("Comment");
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:40]: Once you have timestamp in your datastore then you add to the query to return to you the items in order.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (chenzizhao) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:60]: Oh I didn't know that. Thanks for mentioning. A timestamp has been added to each comment.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (chenzizhao) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Done.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -50,10 +50,17 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    if (newComment.isEmpty()){
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+    }
+    else{
       Entity commentEntity = new Entity("Comment");
       commentEntity.setProperty("content", newComment);
+      commentEntity.setProperty("timestamp", timestamp);
       this.datastore.put(commentEntity);
+      response.sendRedirect("/index.html");
     }
```
Comment (juanalonsocruz) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: The fewer indented lines of code, the easier it is to read code. Typically, when you have

```
if (check for error #1) {
  express that error #1 occurred
} else if (check for error #2) {
  express that error #2 occurred
} else if (check for error #3) {
  express that error #3 occurred
} else {
  handle correct implementation
}
Return
```

It is is easier to read if for every error case you leave "early" (or seems like it leaves early so the reader doesn't have to keep reading further.

```
if (check for error #1) {
  express that error #1 occurred
  return
} 

if (check for error #2) {
  express that error #2 occurred
  return
}

if (check for error #3) {
  express that error #3 occurred
} 

handle correct implementation
Return
```

See a better explanation in the tott I sent in the chat.

**Pull request merged**